www/makefsdata.js

```diff
@@ -1,41 +1,265 @@
-import fs from "fs";
+import path from 'node:path';
+import fs from 'node:fs';
 
-import { execFile } from "child_process";
-import { fileURLToPath } from "url";
+import { fileURLToPath } from 'node:url';
+
+import assert from 'node:assert';
+
+import pako from 'pako';
 
 const __filename = fileURLToPath(import.meta.url);
 
-const root = path.dirname(__filename).replace(path.normalize("www"), "");
+const root = path.dirname(__filename).replace(path.normalize('www'), '');
 const rootwww = path.dirname(__filename);
+const buildPath = path.join(rootwww, 'build');
+
+const fsdataPath = path.normalize(path.join(root, 'lib/httpd/fsdata.c'));
+
+// These are the same content types that are used by the original makefsdata
+const contentTypes = new Map([
+	['html', 'text/html'],
+	['htm', 'text/html'],
+	[
+		'shtml',
+		'text/html\r\nExpires: Fri, 10 Apr 2008 14:00:00 GMT\r\nPragma: no-cache',
+	],
+	[
+		'shtm',
+		'text/html\r\nExpires: Fri, 10 Apr 2008 14:00:00 GMT\r\nPragma: no-cache',
+	],
+	[
+		'ssi',
+		'text/html\r\nExpires: Fri, 10 Apr 2008 14:00:00 GMT\r\nPragma: no-cache',
+	],
+	['gif', 'image/gif'],
+	['png', 'image/png'],
+	['jpg', 'image/jpeg'],
+	['bmp', 'image/bmp'],
+	['ico', 'image/x-icon'],
+	['class', 'application/octet-stream'],
+	['cls', 'application/octet-stream'],
+	['js', 'application/javascript'],
+	['ram', 'application/javascript'],
+	['css', 'text/css'],
+	['swf', 'application/x-shockwave-flash'],
+	['xml', 'text/xml'],
+	['xsl', 'text/xml'],
+	['pdf', 'application/pdf'],
+	['json', 'application/json'],
+]);
+
+const defaultContentType = 'text/plain';
+
+const shtmlExtensions = new Set(['shtml', 'shtm', 'ssi', 'xml', 'json']);
+
+const skipCompressionExtensions = new Set(['png', 'json']);
+
+const serverHeader = 'GP2040-CE';
 
-const fsdata_filename = path.normalize("/lib/httpd/fsdata.c");
-
-function correctincludes() {
-    fs.readFile(root + fsdata_filename, "utf8", function (error, data) {
-        if (data) {
-            let result = data.replace("#include \"lwip/apps/fs.h\"\r\n#include \"lwip/def.h\"", '#include "fsdata.h"');
-        
-            fs.writeFile(root + fsdata_filename, result, "utf8", function (error) {
-                if (error) {
-                    console.error(error);
-                }
-            });
-        } else {
-            if (error) {
-                console.error(error);
-            }
-        }
-    });
+const payloadAlignment = 4;
+const hexBytesPerLine = 16;
+
+function getFiles(dir) {
+	let results = [];
+	const list = fs.readdirSync(dir, { withFileTypes: true });
+	for (const file of list) {
+		file.path = dir + '/' + file.name;
+		if (file.isDirectory()) {
+			results = results.concat(getFiles(file.path));
+		} else if (file.isFile()) {
+			results.push(file.path);
+		}
+	}
+	return results;
+}
+
+function getLowerCaseFileExtension(file) {
+	const ext = file.split('.').pop();
+	return ext ? ext.toLowerCase() : '';
+}
+
+function fixFilenameForC(filename) {
```

Comment:
Unsure how this works, super interesting.
Does this return the character if it's a latin character otherwise "_" ?

Reply:
Basically yes, it converts all characters that are not valid for C identifiers to "_". The original `makefsdata` has a function of the same name that does this as well.

I'll add a clarifying comment.

Reply:
Thanks for explaining, tried it a bunch to get an understanding.
Should have asked in discord though.

A nit is that :
```
else if (c === '_') {
			varName += c;
		}
```
Can be removed as the default would cover it.